### ISSUE TITLE
Add caching to return Python objects to Excel cells

### ISSUE THREAD

**Issue by fzumstein** [ZoomerAnalytics/xlwings#555]

**Comment by ericremoreynolds** [comment]
References: [1](https://zoomeranalytics.uservoice.com/forums/269851-xlwings/suggestions/15495960-support-general-python-object-as-return-type-or-ar) [2] (http://stackoverflow.com/questions/38537566/when-using-xlwings-can-i-return-a-object-to-excel-cell)

**Comment by skasi7** [comment]
I have a implementation of this feature (based on https://www.pyxll.com/docs/examples/objectcache.html but simpler). The problem is this feature requires event handling (that I also implemented) to be efficient.

If you are interested, I can send you more information/examples/PRs.

**Comment by sdementen** [comment]
+1

**Comment by fzumstein** [comment]
sure, please send!

**Comment by audetto** [comment]
This is similar to the issue I recently opened https://github.com/xlwings/xlwings/issues/1541
I work in finance and have see countless poor implementations of this concept which make spreadsheet a nightmare to follow, with issues like: F9 does not work, memory builds up, multiple sheet do not work together.

I think Microsoft should solve it once and for all, but IMHO xlwings is the next best location to do so for python objects.

What I am writing at the moment uses these ideas

- put the object in a global variable (a dictionary)
- keyed by the cell address
- add some visible tweaks to identify as objects, giving the impression that the object has been recomputed

so one could use keys like

object:[filename.xls]Sheet3!B4@10:45:56

where the time is not used in the lookup (because we must delete the object already in the same cell).
optionally, add the type as visual info (not used in lookup)

DataFrame:[filename.xls]Sheet3!B4@10:45:56

Converters are really a good place for this, as they ensure a full separation between object handling and function code (another pitfall of the many bad designs).

**Comment by sdementen** [comment]
You may try the converter `CacheConverter` below that can be used as:
```python
import pandas
import xlwings as xw


@xw.func
@xw.arg("df", pandas.DataFrame)
@xw.ret(CacheConverter)
def write_df2cache(df):
    return dict(my_value=df, my_other_value="45", the_third_value=dict(info=len(df),cols=df.columns))



@xw.func
@xw.arg("df", CacheConverter)
def read_cache(df):
    return df


@xw.func
def see_cached_keys():
    return [[e] for e in _cache_name_timestamp]
```
In Excel, you use `write_df2cache` to read a range. This will return 3 values as reference similar to:
_my_value@Sheet1!$E$7[2022-02-06   09:55:11.750726]{DataFrame-3}
my_other_value@Sheet1!$E$7[2022-02-06   09:55:11.750726]{str-2}
the_third_value@Sheet1!$E$7[2022-02-06   09:55:11.750726]{dict-2}_

You can thereafter extract the values behind one of the 3 references by using `read_cache` with as argument one of the 3 references.

The API is not final but could be a good start for discussion.


```python
from xlwings.conversion import Converter

_cache_name_timestamp = {}


class CacheConverter(Converter):
    """Convert an object to a reference. The reference can include the 'caller' address to ensure
    uniqueness (not needed if singleton - caller=False).

    The object are stored in a global dict by their key and looked up when reading.
    
    When used on a return value, the converter will transform the value into a reference. 
    If the value is a dict, it will return a reference for each item in the dict.
    
    When used in an input, the converter will return the value that was stored in a reference.
    """

    @staticmethod
    def read_value(value, options):
        if not value:
            return
        key = value.split("[")[0]
        if key in _cache_name_timestamp:
            return _cache_name_timestamp[key]
        else:
            raise ValueError(f"No data available for {key}")

    @staticmethod
    def write_value(value, options):
        if value is None:
            return
        now = datetime.datetime.today()
        if options.get("caller", True):
            caller = xw.apps.active.api.Caller
            caller = f"@{caller.Worksheet.Name}!{caller.Address}"
        else:
            caller = ""

        if not isinstance(value, dict):
            if not caller:
                raise ValueError("You cannot use 'caller'=False if the function does not return a dict.")
            value = dict(anonymous=value)

        hashes = []
        for name, item in value.items():
            key = f"{name}{caller}"
            _cache_name_timestamp[key] = item
            hashes.append(f"{key}[{now}]{{{type(item).__name__}-{len(item)}}}")

        if options.get("vertical", True):
            hashes = [[h] for h in hashes]

        return hashes
```

**Comment by sdementen** [comment]
Hi @fzumstein 
Do you think my converter could be added to xlwings or to specific?

**Comment by fzumstein** [comment]
Yeah I guess we still want to add that object caching functionally at some point, it just hasn't made it to the top...

**Comment by fzumstein** [comment]
I know it's been a while, but this has finally been implemented in [xlwings Server](https://github.com/xlwings/xlwings-server) as Office.js add-ins allowed for proper implementation, pretty much satisfying all of @audetto's requirements 😎 You can have a look at the docs: https://docs.xlwings.org/en/latest/pro/server/officejs_custom_functions.html#object-handles or here's a screenshot:

![image](https://github.com/user-attachments/assets/50e3db4a-3b95-476e-b07c-f0ec10b88ea5)

**Comment by MarcPelland** [comment]
Is it possible to add this to the local version of xlwings?

**Comment by fzumstein** [comment]
Yes, but you can't get the icon/card and you won't get the green double-arrows when it's recalculating. You can have a look at the implementation for inspiration, but for said differences and because xlwings Server is a shared server, the code would have to be adopted quite a bit. I also don't currently have the time to implement this myself in the OSS version.

**Comment by MarcPelland** [comment]
Thanks for your response and I fully understand. From my further research I was able to find that this is implemented in xloil (https://github.com/cunnane/xloil). I'm just mentioning it here in case it saves someone a lot of time looking.
